lib/srv/app/server.go

```diff
@@ -592,29 +601,112 @@ func (s *Server) HandleConnection(conn net.Conn) {
 }
 
 func (s *Server) handleConnection(conn net.Conn) error {
+	// Make sure everything here is wrapped in the tracking read connection for monitoring.
+	ctx, cancel := context.WithCancel(s.closeContext)
+	tc, err := srv.NewTrackingReadConn(srv.TrackingReadConnConfig{
+		Conn:    conn,
+		Clock:   s.c.Clock,
+		Context: ctx,
+		Cancel:  cancel,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
 	// Proxy sends a X.509 client certificate to pass identity information,
 	// extract it and run authorization checks on it.
-	tlsConn, user, app, err := s.getConnectionInfo(s.closeContext, conn)
+	tlsConn, user, app, err := s.getConnectionInfo(ctx, tc)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	authCtx, _, err := s.authorizeContext(context.WithValue(ctx, auth.ContextUser, user))
+	if err != nil {
+		// If the app is not TCP, write an HTTP error response and return nil instead of the error.
+		// This will ensure that HTTP clients will see a "403 Forbidden" code and message and not a 500.
+		if !app.IsTCP() {
+			code := trace.ErrorToCode(err)
+			resp := http.Response{
+				StatusCode: code,
+				Body:       io.NopCloser(bytes.NewReader([]byte(http.StatusText(code)))),
+			}
+			if writeErr := resp.Write(tlsConn); writeErr != nil {
+				s.log.Warnf("error writing HTTP response: %v", writeErr)
+			}
+
+			// Since we're not passing this connection to the http server, we should close
+			// the connection explicitly here.
+			if closeErr := conn.Close(); closeErr != nil {
+				s.log.Warnf("error closing connection: %v", closeErr)
+			}
+
+			return nil
+		}
+
+		return trace.Wrap(err)
+	}
+
+	err = s.monitorConn(ctx, tc, authCtx)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	// Application access supports plain TCP connections which are handled
 	// differently than HTTP requests from web apps.
 	if app.IsTCP() {
-		return s.handleTCPApp(s.closeContext, tlsConn, user, app)
+		identity := authCtx.Identity.GetIdentity()
+		return s.handleTCPApp(ctx, tlsConn, &identity, app)
```

Comment:
should we use `tc` instead of `tlsConn`?

Reply:
So tlsConn uses `tc` underneath -- shouldn't this work? Using `tc` directly was not working for me, FWIW.

As for your other comment about serving the original connection to `handleHTTPApp`, I'm slightly worried about a race condition where:

1. Monitor creation fails.
2. Run `handleHTTPApp`.
3. The lock is deleted in that time.
4. The `handleHTTPApp` unexpectedly works, creating a connection with no monitoring.

Am I overthinking?

Reply:
my bad, i missed the part `getConnectionInfo` is using `tc`,

Reply:
No worries! I'm still figuring this code out, it's slightly byzantine.